viewing_party/party.py

```diff
@@ -1,23 +1,248 @@
+    ratings_lst = []
+    #this gives a list of ratings
+    for movie in user_data["watched"]:
+        ratings_lst.append(movie["rating"])
+
+
+    #adding an if to deal with if it's an empty list:
+    if len(ratings_lst) > 0:
+        #this calculates the average:
+        avg_rating = sum(ratings_lst) / len(ratings_lst)
+        return avg_rating
+    else: 
+        return 0
+
+
+def get_most_watched_genre(user_data):
+    #This will make a dictionary, and add a count of how many times each genre appears. 
+    genre_count = {}
+    #keys will be genres (strings)
+    #values will be the number of movies in that genre.
+    #adding in an if to deal with if watched is an empty list:
+    if len(user_data["watched"]) > 0:
+        for movie in user_data["watched"]:
+            genre = movie["genre"]
+            if genre not in genre_count:
+                genre_count[genre] = 1
+            else:
+                genre_count[genre] += 1
+        #now, genre_count is complete.  Let's return which key has the highest value associated with it.
+        most_watched_genre = max(genre_count, key = genre_count.get)
+        return most_watched_genre
+    else:
+        return None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    #takes in user_data (a dictionary with 3 keys, watched and watchlist and friends)
+    #now, user_data has a third key: "friends".
+    #the value associated with this is a list of dictionaries, one dictionary per friend.
+    # The dictionary for each friend has "watched", with a list of movies. (which is itself a dict, remember.) 
+    #This function outputs the list of movies (dicts) in watched that are unique, and not watched by another friend.
+    
+    user_watched_titles = get_user_watched_titles(user_data)
+    friend_watched_titles = get_friends_watched_titles(user_data)
+
+    #set difference
+    unique_watched_titles = user_watched_titles - friend_watched_titles
+
+    #retrieve the dictionary entries from the original user_data
+    unique_watched_movies = []
+    for movie in user_data["watched"]:
+        if movie["title"] in unique_watched_titles:
+            unique_watched_movies.append(movie)
+            unique_watched_titles.remove(movie["title"]) #remove it to prevent duplicates
+    return unique_watched_movies
+
+def get_user_watched_titles(user_data):
+    #returns a set of titles user has watched.
+    user_watched_titles = []
+    for movie in user_data['watched']:
+        user_watched_titles.append(movie["title"])
+    return set(user_watched_titles)
+
+def get_friends_watched_titles(user_data):
+    #returns a set of titles of movies friend has watched.
+    friend_watched_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_titles.append(movie["title"])

```

Comment:
Curious if instead of initializing these as lists, could you initialize them as sets and use .add()? This prevents the conversion to a set which eats at the time complexity (O(n) where n is the length of the list to be converted to a set). Otherwise very helpful helper functions!